grpc/binlog/v1alpha/binarylog.proto

```diff
@@ -31,8 +31,14 @@ message GrpcLogEntry {
     SEND_TRAILING_METADATA = 2;
     SEND_MESSAGE = 3;
     RECV_INITIAL_METADATA = 4;
+    // On server side, this event means the client has half-closed.
+    // On client side, this is means the RPC has now ended.
     RECV_TRAILING_METADATA = 5;
     RECV_MESSAGE = 6;
+    // Client side only.
+    CLIENT_SEND_CANCEL = 7;
+    // Server side only. This means the RPC has now ended.
+    SERVER_END_CALL = 8;
```

Comment:
It looks like this is conflating two cases:

1. The server receives a RSET_STREAM from the client after it has sent trailing metadata.  In this case, we probably don't need to log the event in the first place, because the server should consider the call complete once it sends trailing metadata.

2. The server receives a RSET_STREAM from the client before it has sent trailing metadata.  In this case, we could record the event as SERVER_RECV_CANCEL.

I think this representation would be simpler than SERVER_END_CALL, where you need to look at a separate value to understand what happened.

Reply:
@ejona86 can you give some guidance here since you've been looking into the gRPC call semantics

Reply:
For (1) I would caution you to realize there's the _service_ and the _grpc implementation_. The _service_ may have completed with trailers (which is SEND_TRAILING_METADATA), but the RST_STREAM (or I/O failure, etc) may arrive before that is sent on the wire. Note this is easy to trigger, because trailers are limited by flow control and thus are client-controlled on how long they take to send.

For (1), if the grpc implementation has actually sent the trailers then there would have been a ServerCallEnded{NORMAL}. Then receiving a RST_STREAM does nothing and wouldn't be logged at all, because the call is already complete.

The "recv" of SERVER_RECV_CANCEL isn't quite right, as the server may have triggered it itself, as is the case with Deadlines. Even an I/O failure seems a bit weird to call it SERVER_RECV_CANCEL. Now simply SERVER_CANCEL could be okay, but that's basically the same as ServerCallEnded{ABNORMAL}. The question around ServerCallEnded really seems to be whether ServerCallEnded{NORMAL} is a "thing."

Reply:
I'm looking at this from the context of the analyzer tooling, and there is an existing feature that goes through a log and reports on the current in flight RPCs. Having an explicit ServerCallEnded{NORMAL} sentinel would make that sort of thing much easier.

Reply:
Since we aren't thinking of using the same tools any longer, I don't see any use for replicating features.

I agree that it might be useful to know how the call ended - whether it was cancelled or something else. For this, I suggest that we log the status (the status code and string). Logging the status would be an indication that the call has been completed (doesn't matter whether it completed properly or cancelled or some other error). This also covers the case of SEND_CANCEL, RECV_CANCEL, etc. and also follows the idea that the filter will be treated as something closer to the application and everything it does is something that the users could have done themselves.

Reply:
Logging the status that the server app sends is sufficient for normal terminations, but if the client cancels the RPC then no status or trailing metadata will be sent. I'm thinking of something like this:
1. client sends initial metaata
1. server receives client initial metadata
1. sever sends initial metadata
1. client sends cancel (RST_STREAM)
1. server receives cancel

The RPC is now done but no status was sent by the server application. Perhaps the C API uses status to express this? For Java it seems misleading to record that the call ended with a Status.CANCELLED.

Another case is the server can send out a Status.CANCELLED, but because it is a normal termination, the RPC should be recorded as ServerCallEnded{NORMAL}.

RE: tools
I agree we don't need to replicate features exactly but being able to report on the liveness of an RPC  seems like a basic use case.

Reply:
@zpencer Why would it be misleading? The RPC was cancelled. As to why it was cancelled, the details could be found in the message. For a cancel originating from the client side, it would be something like - Received RST_STREAM
As for the second case, why do we need to call that 'normal' termination? I would imagine the user to expect normal to be GRPC_STATUS_OK. For the case that you describe it should suffice to provide Cancelled status with the message that the user provided.

I am not against logging SEND_CANCEL which can be from either the server or the client, but it just seems redundant information. SEND_CANCEL does not seem to be adding anything new IMO.

The other alternative would be to have a SEND_CANCEL type and then a type to log the status code on the client CLIENT_STATUS_CODE, while on the server, it would just be a boolean of -
- true if the server perceives the call as terminated with Status::OK 
- false if it is anything else (includes server cancellation, client cancellation and all other errors)
But, in this approach, we are losing data on why it was not an OK status on the server

Reply:
As for the liveness of the RPC, the final status on the RPC would denote that the RPC has ended.

Reply:
For the first example: Are you suggesting that the server should log a SEND_TRAILING_METADATA message when nothing was sent? Based on the liveness comment it sounds to be the case. No status or metadata was sent, which is why it's misleading. I see two options: 1) lie and claim there was a SEND_TRAILING_METADATA when in fact there wasn't anything sent, or 2) we have some other type of event like ServerCallEnded{ABNORMAL}.

For the second case: normal vs abnormal refers to whether the call terminated by the server sending a Status+Metadata. So I wouldn't say it's redundant information.

Reply:
I am not suggesting claiming that there was a send_trailing_metadata. I am proposing separating the notion of status code at the server side from what was actually sent to indicating how it came to an end. This means that the status code observed at the server side and that observed at the client side can potentially be different for error cases. This would not mean that the server actually sent a trailing metadata with that status code. It's just how the RPC ended according to the server. Note that for the cases where the server actually sends the status in trailing metadata which is the normal behavior in your proposal, the status code would match.

Also, since we have a separate type for metadata which we log, it would be unambiguous whether metadata was sent or not. We just need to know when and how the call was terminated even if trailing metadata was not sent.

Reply:
Java has a similar sort of "server side close status" today, except it is an internal mechanism. Does the C API expose this status to applications?